File: playwright-tests/src/pages/ContactPage.ts
import { expect, Locator, Page } from '@playwright/test';

export class ContactPage {
  readonly heading: Locator;

  constructor(private readonly page: Page) {
    this.heading = page.getByRole('heading', { name: 'Send Us a Message' });
  }

  async goto(): Promise<void> {
    await this.page.goto('/#contact');
    await expect(this.heading, 'Contact form heading should be visible').toBeVisible();
  }

  async submitMessage(data: ContactMessage): Promise<void> {
    await this.fillMessage(data);
    await this.submit();
  }

  async fillMessage(data: ContactMessage): Promise<void> {
    await this.page.getByRole('textbox', { name: 'Name' }).fill(data.name);
    await this.page.getByRole('textbox', { name: 'Email' }).fill(data.email);
    await this.page.getByRole('textbox', { name: 'Phone' }).fill(data.phone);
    await this.page.getByRole('textbox', { name: 'Subject' }).fill(data.subject);

    // The message textarea has no accessible name in the current public UI.
    await expect(this.messageInput, 'Message textarea should be visible').toBeVisible();
    await this.messageInput.fill(data.message);
  }

  async submit(): Promise<void> {
    await expect(this.submitButton, 'Contact Submit button should be enabled').toBeEnabled();
    await this.submitButton.click();
  }

  async expectSuccessMessage(): Promise<void> {
    await expect(
      this.page.getByText(/thanks|thank you|message sent/i).first(),
      'Contact form should show a success acknowledgement',
    ).toBeVisible();
  }

  async expectValidationFeedback(): Promise<void> {
    const validationErrors = this.page.locator('.alert-danger, .invalid-feedback, [role="alert"]').filter({
      hasText: /must|valid|error|required|between|phone|email/i,
    });
    await expect(validationErrors.first(), 'Invalid contact submission should show validation feedback').toBeVisible();
  }

  private get submitButton(): Locator {
    return this.page.getByRole('button', { name: 'Submit' });
  }

  private get messageInput(): Locator {
    return this.page.locator('textarea').first();
  }
}

export type ContactMessage = {
  name: string;
  email: string;
  phone: string;
  subject: string;
  message: string;
};

File: selenium-web-tests/src/test/java/com/portfolio/pages/ContactPage.java
package com.portfolio.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class ContactPage extends BasePage {
    private static final By HEADING = By.xpath(
            "//*[self::h1 or self::h2 or self::h3][normalize-space()='Send Us a Message']"
    );
    private static final By NAME_INPUT = By.cssSelector("#name");
    private static final By EMAIL_INPUT = By.cssSelector("#email");
    private static final By PHONE_INPUT = By.cssSelector("#phone");
    private static final By SUBJECT_INPUT = By.cssSelector("#subject");
    private static final By MESSAGE_INPUT = By.cssSelector("#description");
    private static final By SUBMIT_BUTTON = By.xpath("//button[normalize-space()='Submit']");
    private static final By SUCCESS_MESSAGE = By.xpath(
            "//*[contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'thank you') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'thanks') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'message sent')]"
    );
    private static final By VALIDATION_FEEDBACK = By.xpath(
            "//*[contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'must') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'valid') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'error') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'required') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'between') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'phone') or " +
                    "contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'email')]"
    );

    public ContactPage(WebDriver driver, String baseUrl) {
        super(driver, baseUrl);
    }

    public void gotoPage() {
        open("/#contact");
        visible(HEADING);
    }

    public void fillMessage(String name, String email, String phone, String subject, String message) {
        type(NAME_INPUT, name);
        type(EMAIL_INPUT, email);
        type(PHONE_INPUT, phone);
        type(SUBJECT_INPUT, subject);
        type(MESSAGE_INPUT, message);
    }

    public void submit() {
        click(SUBMIT_BUTTON);
    }

    public void expectSuccessMessage() {
        visible(SUCCESS_MESSAGE);
    }
}
